src/include/baselib/jni/JavaVirtualMachine.h

```diff
@@ -0,0 +1,178 @@
+                    );
+
+                bl::fs::path jvmPath( *javaHome );
+
+                if( bl::os::onWindows() )
+                {
+                    jvmPath += "/jre/bin/server/jvm.dll";
+                }
+                else
+                {
+                    jvmPath += "/jre/lib/amd64/server/libjvm.so";
+                }
+
+                jvmPath = bl::fs::normalize( jvmPath );
+
+                BL_CHK_USER_FRIENDLY(
+                    false,
+                    bl::fs::exists( jvmPath ),
+                    BL_MSG()
+                        << "Path doesn't exist "
+                        << bl::fs::normalizePathParameterForPrint( jvmPath )
+                    );
+
+                if( bl::os::onWindows() )
+                {
+                    /*
+                     * Remove "\\?\" prefix from JavaVM path on Windows.
+                     * Otherwise after the JVM library is loaded the call to JNI_CreateJavaVM will fail with:
+                     *     Error occurred during initialization of VM
+                     *     java/lang/NoClassDefFoundError: java/lang/Object
+                     */
+
+                    return bl::fs::detail::WinLfnUtils::chk2RemovePrefix( std::move( jvmPath ) ).string();
+                }
+                else
+                {
+                    return jvmPath.string();
+                }
+            }
+
+            void createJvm()
+            {
+                BL_MUTEX_GUARD( g_lock );
+
+                /*
+                 * Creation of multiple VMs in a single process is not supported.
+                 * http://docs.oracle.com/javase/8/docs/technotes/guides/jni/spec/invocation.html#JNI_CreateJavaVM
+                 */
+
+                BL_CHK_USER_FRIENDLY(
+                    true,
+                    g_javaVMCreated,
+                    "JavaVM has already been created"
+                    );
+
+                const auto procAddress = bl::os::getProcAddress( m_jvmLibrary, "JNI_CreateJavaVM" );
+
+                const auto jniCreateJavaVM = reinterpret_cast< jint ( JNICALL* )( JavaVM**, void**, void *) >( procAddress );
+
+                JavaVMInitArgs vmArgs;
+                vmArgs.version = JNI_VERSION_1_8;
+                vmArgs.nOptions = 0;
+                vmArgs.options = nullptr;
+                vmArgs.ignoreUnrecognized = JNI_FALSE;
+
+                BL_LOG(
+                    bl::Logging::debug(),
+                    "Creating JVM"
+                    );
+
+                JNIEnv *jniEnv;
+
+                jint jniReturn = jniCreateJavaVM(
+                    &m_javaVM,
+                    ( void** )&jniEnv,
+                    &vmArgs
+                    );
+
+                BL_CHK_USER_FRIENDLY(
+                    true,
+                    jniReturn != JNI_OK,
+                    BL_MSG()
+                        << "Failed to create JVM. Error code "
+                        << jniReturn
```

Comment:
Can you obtain more detailed error handling information from the JNI layer (other than error code)? E.g. error message? Probably also best to define new exception type (e.g. JvmException) to throw